### ISSUE TITLE
Include in-memory (sync match) backlog in `ApproximateBacklogCount`

### ISSUE THREAD

**Issue by bergundy** [temporalio/temporal#6760]
There are some tasks that will never go to the durable backlog, such as queries and Nexus tasks.
In those cases, and in cases where there's a high sync match rate, it's useful to know how many tasks are in the sync match backlog in order to autoscale workers.

**Comment by sivagirish81** [comment]
I’d like to propose an implementation approach for this and get maintainer feedback before opening a PR.

My understanding is that the gap is specifically around non-durable sync-match tasks, such as Queries and Nexus tasks. These can create real pressure on a Task Queue while the durable backlog remains `0`, because they may only be waiting in Matching memory for a poller. In that case, autoscalers using `ApproximateBacklogCount` can incorrectly conclude that there is no pending work and avoid scaling Workers up.

```text
ApproximateBacklogCount =
  durable backlog count
  + in-memory Query/Nexus sync-match waiting count
```

The idea is to count Query/Nexus tasks while they are waiting in Matching’s sync-match path.

The proposed behavior is:

- Count only Query/Nexus tasks that are actively waiting for a poller in Matching memory.
- Increment when the task enters the in-memory wait path.
- Decrement when the task leaves that wait path, including:
  - matched to a poller
  - forwarded
  - timed out or canceled
  - reprocessed/evicted due to task queue/user data changes
- Drop the count at dispatch/match time, not when the Worker finishes handling the Query/Nexus task, since backlog should mean “waiting for a Worker,” not “currently executing.”
- Avoid double-counting across forwarding/partition paths by counting the task only in the matcher where it is actually waiting.

For the newer matcher path, I think the cleanest place is the `taskPQ` lifecycle.
Since the newer matcher keeps waiting tasks in `taskPQ`, tracking at push/remove/pop time should naturally cover the important exits:

```text
enters sync-match wait   -> added to taskPQ       -> increment
matched to a poller      -> removed from taskPQ -> decrement
context timeout/cancel   -> removed from taskPQ -> decrement
reprocessed/evicted      -> removed from taskPQ -> decrement
still waiting            -> remains in taskPQ   -> still counted
```

For the classic matcher path, there is no equivalent `taskPQ`, so I’d track around the local sync-offer wait in:

- `service/matching/matcher.go`
  - `OfferQuery`
  - `OfferNexusTask`
  - shared `syncOfferTask`

One subtlety there is forwarding: 
The classic matcher has no equivalent taskPQ like the newMatcher. Query and Nexus tasks wait by blocking inside syncOfferTask on a local channel send. Therefore the tracker must bracket only that local wait. If the forwarding branch wins, the task must be decremented before ForwardQueryTask/ForwardNexusTask, because it is no longer waiting for a local poller. If forwarding is rate-limited and syncOfferTask falls back to local waiting, the task can be counted again for that next local wait episode.

Then I’d expose matcher-owned sync backlog stats, for example:

```go
SyncMatchStatsByPriority() map[int32]*taskqueuepb.TaskQueueStats
```

and merge those into the existing stats path in:

- `service/matching/physical_task_queue_manager.go`
  - `physicalTaskQueueManager.GetStatsByPriority`

That keeps `task_queue_partition_manager.go` consuming the already-merged `TaskQueueStatsByPriority`, so `TaskQueueStats`, `TaskQueueStatsByPriorityKey`, Describe output, and metrics stay consistent.

The resulting flow would be:

```text
classic matcher sync-offer tracker
        +
new matcher taskPQ sync-wait tracker
        ↓
matcher sync backlog stats by priority
        ↓
physicalTaskQueueManager.GetStatsByPriority()
        ↓
TaskQueueStatsByPriorityKey
        ↓
aggregate task queue stats
        ↓
TaskQueueStats.ApproximateBacklogCount
```

For `ApproximateBacklogAge`, I think it should remain a unified public value:

```text
ApproximateBacklogAge =
  max(oldest durable waiting age, oldest sync-match waiting age)
```

That keeps the autoscaler-facing meaning simple:

```text
count = how many tasks are waiting for a Worker
age   = how long the oldest waiting task has waited
```

Does this sound like the right place to track and merge the in-memory sync-match backlog, or would maintainers prefer this to live somewhere else in Matching?

**Comment by dnr** [comment]
Right.. for normal activity/workflow tasks, there shouldn't be tasks blocked in memory, at least not for very long, so we can focus on query/nexus.

Let's just leave out classic matcher, I think. We're going to remove it soon, and that'll keep things simpler.

I see two general approaches:

1. What you suggested, counting in `taskPQ`. It could look at `source` to count only tasks not from a backlog and not forwarded. Except, when a task is forwarded it's removed from `taskPQ`, so actually a task should only be sitting in one or another. So maybe we just count all `source != backlog` and sum across partitions.

2. Count at a higher level, e.g. in matching engine when we add to the in-memory response channel maps.

Given the places we want to add in these numbers, I agree (1) makes more sense. So yeah, what you suggested makes sense. I'd just switch to using `source` and ignoring classic matcher.

There may be some complication around the "logical" vs "physical" backlog. @Shivs11  may have a comment here.

@bergundy  as the original requester, any thoughts in the past year or does this still make sense?
